### PR TITLE
FRON-43 Implement endpoint for statistics

### PR DESCRIPTION
**Pull request by MCFreddie777**

### CODE REVIEW

File: antiplag/models.py [hunk 1]
```diff
@@ -17,6 +19,24 @@ class SubmissionStatus(models.TextChoices):
...
+        if cached_value is not None:
+            return cached_value
+        else:
+            result = cls.objects.filter(
+                status=cls.SubmissionStatus.PROCESSED
+            ).aggregate(
+                average_time=models.Avg(models.F("updated_at") - models.F("created_at"))
+            )
+            return ceil(result.get("average_time").seconds / 60)  # in minutes
+
```
Comment (OliverKanat) [inline antiplag/models.py:39]: Am I missing something? Don't you need to set the cache as well?

File: nlp/elastic.py [hunk 2]
```diff
@@ -1,8 +1,14 @@
+from django.core.cache import cache
 from elasticsearch_dsl import Search
 from elasticsearch_dsl.query import MoreLikeThis
 
 
 class Elastic:
+    @staticmethod
+    def count():
+        cached_value = cache.get("corpus_size", None)
+        return cached_value if cached_value else Search(index="documents").count()
```
Comment (OliverKanat) [inline nlp/elastic.py:10]: Same here for setting the cache.

File: antiplag/models.py [hunk 3]
```diff
@@ -17,6 +19,24 @@ class SubmissionStatus(models.TextChoices):
...
         return f"Submission {self.id} ({self.status})"
 
+    @classmethod
+    def count(cls):
+        cached_value = cache.get("submission_count", None)
+        return cached_value if cached_value else cls.objects.count()
+
+    @classmethod
+    def average_time(cls):
+        cached_value = cache.get("submission_avg_time")
```
Comment (OliverKanat) [inline antiplag/models.py:29]: Why are you using `.get()` without the fallback argument? Method higher it is used with the fallback argument.

`cache.get(key)` returns `None` as default, if no cached item is found, but I think specifying `None` makes it a bit more readable.

File: antiplag/views.py [hunk 4]
```diff
@@ -3,15 +3,25 @@
...
+
+class Stats(APIView):
+    def get(self, request):
+        return Response(
+            data={
+                "submission_count": Submission.count(),
+                "submission_avg_time": Submission.average_time(),
+                "corpus_size": Elastic.count(),
+            }
+        )
```
Comment (OliverKanat) [inline antiplag/views.py:24]: Maybe it would be better to cache the view and don't deal with caching every stat? What do you think?

**Review by OliverKanat — CHANGES_REQUESTED** [review]

File: antiplag/views.py (hunk 4 above)
Comment (MCFreddie777) [reply antiplag/views.py:24]: Might do.

File: antiplag/models.py (hunk 3 above)
Comment (MCFreddie777) [reply antiplag/models.py:29]: Yes I know, previously I had it implicitly, but I like it explicit - therefore I removed checks for `X is not None` and replaced with `not X`, but I forgot to replace this part as well. 

Though, this discussion is irrelevant, I'm going to refactor it to view caching as you suggested.

File: antiplag/models.py (hunk 1 above)
Comment (MCFreddie777) [reply antiplag/models.py:39]: image

File: antiplag/views.py [hunk 5]
```diff
@@ -3,15 +3,28 @@
...
 from .constants import TEXT_SUBMISSION_NAME, CONTENT_TYPE_FILE, CONTENT_TYPE_TEXT
 from .tasks import process_documents
+from nlp.elastic import Elastic
 
-from django.conf import settings
-from nlp.text_preprocessing import preprocess_text
-from django.utils.translation import ugettext as _
+
+class Stats(APIView):
+    @method_decorator(cache_page(60 * 60), name="dispatch")
```
Comment (OliverKanat) [inline antiplag/views.py:19]: I don't think you used `method_decorator` correctly. Either decorate class and pass function name (of function you want to decorate) to `name` arg or keep it as is and remove `name` arg.

Comment (OliverKanat) [reply antiplag/views.py:19]: https://docs.djangoproject.com/en/3.1/topics/class-based-views/intro/#decorating-the-class

**Review by OliverKanat — CHANGES_REQUESTED** [review]

File: antiplag/views.py (hunk 5 above)
Comment (MCFreddie777) [reply antiplag/views.py:19]: yeah, seems like it I got confused and combined two ways. Not a big fan of decorators, to be honest - seems like a syntactic sugar and black box to me.

Mostly now, when developing docker, my PyCharm setup doesn't recognize the packages and I can't explore the functions from python using IntelliSense.

**Review by OliverKanat — APPROVED** [review]

**Review by Scr0ll — APPROVED** [review]

**Pull request merged**